repo/content/index/id.go

```diff
@@ -94,6 +96,45 @@ func (i ID) less(other ID) bool {
 	return bytes.Compare(i.data[:i.idLen], other.data[:other.idLen]) < 0
 }
 
+// LoggableID is a byte array that stores UTF-8 characters of content ID for logging purposes
+// without incurring allocation cost.
+type LoggableID [2*hashing.MaxHashSize + 2]byte
+
+// String returns string representation of content ID that's only valid as long as LoggableID is
+// not garbage collected. Use with caution.
+func (l LoggableID) String() (s string) {
+	// last byte encodes valid length
+	length := int(l[2*hashing.MaxHashSize+1])
+	if length == 0 {
+		return ""
+	}
+
+	// last byte encodes length
+	// *reflect.StringHeader can't be constructed, so we refer to output variable here
+	shdr := (*reflect.StringHeader)(unsafe.Pointer(&s)) //nolint:gosec
+	shdr.Data = uintptr(unsafe.Pointer(&l))             //nolint:gosec
+	shdr.Len = length
+
+	return
+}
+
+// ForLogging returns short-term LoggableID suitable for logging that avoids memory allocation.
+func (i ID) ForLogging() LoggableID {
+	var l LoggableID
```

Comment:
While this does not allocate a string, wouldn't this end up allocating a `LoggableID` since it is escaping the context. Is the compiler able to optimize it?

Reply:
rewrote the PR to use explicit buffer pool and allocation free `Append()` pattern.